src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java

```diff
@@ -135,4 +109,126 @@ public static void assertSamValid(final File sam) throws IOException {
+            return "name different read1:" + read1.getReadName() + " read2:" + read2.getReadName() ;
+        }
+        if (!Objects.equals(read1.getFlags(), read2.getFlags())){
+            return "flags different read1:" + read1.getReadName() + " read2:" + read2.getReadName() ;
+        }
+        if (!Objects.equals(read1.getInferredInsertSize(), read2.getInferredInsertSize())){
+            return "getInferredInsertSize different read1:" + read1.getReadName() + " read2:" + read2.getReadName() + " (" + read1.getInferredInsertSize()+ " vs " + read2.getInferredInsertSize() + ")";
+        }
+        if (!Objects.equals(read1.getMappingQuality(), read2.getMappingQuality())){
+            return "getMappingQuality different read1:" + read1.getReadName() + " read2:" + read2.getReadName() ;
+        }
+        //Note the comparator does check flags, MQ and insert size but we want better messages
+        if (0 != coordinateComparator.fileOrderCompare(read1, read2)){
+            return "file order read1:" + read1.getContig() + ":" + read1.getAlignmentStart() + " read2:" + read2.getContig() + ":" + read2.getAlignmentStart();
+        }
+        if (!Objects.equals(read1.getCigar(), read2.getCigar())){
+            return "getCigar different read1:" + read1.getReadName() + " read2:" + read2.getReadName() ;
+        }
+        if (!Arrays.equals(read1.getReadBases(), read2.getReadBases())){
+            return "getReadBases different read1:" + read1.getReadName() + " read2:" + read2.getReadName() ;
+        }
+        if (!Arrays.equals(read1.getBaseQualities(), read2.getBaseQualities())){
+            return "getBaseQualities different read1:" + read1.getReadName() + " read2:" + read2.getReadName() ;
+        }
+        final List<SAMRecord.SAMTagAndValue> attr1 = read1.getAttributes();
+        final List<SAMRecord.SAMTagAndValue> attr2 = read2.getAttributes();
+        if (attr1.size() > attr2.size()){
+            return "lost attributes read1:" + read1.getReadName() + " read2:" + read2.getReadName() + "("+ attr1.size() + " vs " + attr2.size() +")" ;
+        }
+
+        //We want to compare attributes regardless of order, so we sort first attributes by name
+        final SortedMap<String, Object> values1 = new TreeMap<>();
+        final SortedMap<String, Object> values2 = new TreeMap<>();
+
+        for (SAMRecord.SAMTagAndValue samTagAndValue : attr1) {
+            values1.put(samTagAndValue.tag, samTagAndValue.value);
+        }
+        for (SAMRecord.SAMTagAndValue samTagAndValue : attr2) {
+            values2.put(samTagAndValue.tag, samTagAndValue.value);
+        }
+
+        for (int i = 0; i < attr1.size(); i++) {
+            final String tag = attr1.get(i).tag;
+            if (!values2.containsKey(tag)){
+                return "attribute " + tag + " is missing from read2:" + read2.getReadName() ;
+            }
+            final Object v1 = values1.get(tag);
+            final Object v2 = values2.get(tag);
```

Comment:
Instead of calling both `containsKey()` and `get()` on `values2`, just call `get()` and check the return value for null (saves one additional hash lookup per attribute).